### PR TITLE
refactor openfoam packages

### PR DESCRIPTION
**Pull request by olesenm**
use common OpenfoamArch class to manage common openfoam naming and compiler handling.
Only carp about unsupported arch/compiler at the build stage, which should improve robustness.
export special-purpose FOAM_PROJECT_DIR env for dependent packages, which can be used to locate the etc/bashrc file.
- fixup handling of make_jobs (now use global).

### CODE REVIEW

**Comment by adamjstewart** [comment]
@olesenm I promise I'm not ignoring your OpenFOAM PRs. There's just a lot to review here and I'm not very familiar with OpenFOAM. Most Spack packages are 5-10 lines long. I assume there's no way to convince the various OpenFOAM developers to switch to CMake or Autotools? 😆 

One thing worth doing would be to include all of the variables set in `$FOAM_PROJECT_DIR/etc/bashrc` in the resulting module file. Right now, if you want to use OpenFOAM, you have to run:
```
$ module load openfoam-org
$ source "$FOAM_PROJECT_DIR/etc/bashrc"
```
If you add the appropriate environment variables to the module file, you won't need to run the second command anymore. See `intel-parallel-studio` for a similarly complex example of how to set environment variables that get added to the generated module file.

**Comment by olesenm** [comment]
I have the same concern about environment variables, but I want to leverage the normal OpenFOAM bashrc information and not hardcode the equivalent logic inside of the package. Otherwise I greatly fear that the environment will very quickly become out of sync with what the bashrc logic produces.
Unfortunately using EnvironmentModifications.from_sourcing_files() bombs out. Besides which, it isn't flexible enough. We'd need to source the bashrc (for/from stage.source_path), replace the prefix/project directories with their final locations. Purge a few entries and/or replace them with more generalized values (partly using the user's environment). Example,

    setenv FOAM_USER_APPBIN  "$env(HOME)/$env(USER)-$WM_PROJECT_VERSION/..."

This unfortunately also needs to go back into the LD_LIBRARY_PATH and it doesn't look like the current spack tools will be useable here. I have an alternative (hack) to do the same thing and convert the environment changes to a ModuleInclude.tcl, but no obvious way to integrate this into the spack install, nor if this is even a sane approach in the first place.

**Comment by adamjstewart** [comment]
> Unfortunately using `EnvironmentModifications.from_sourcing_files()` bombs out.

Yeah, I've seen the same problem. The problem is that `setup_environment` runs before the installation, so there's no way to add it to your module file post-installation. @alalazo I wonder if we should run `setup_environment` pre-installation to set `spack_env` and then run `setup_environment` again post-installation to add `run_env` to the module file. Or create two separate functions. The same problem affects the `intel-parallel-studio` package, which also comes with a handy bashrc that we have manually added to Spack. As development of these projects continues, these environment variables will change and become trickier to manage.

File: var/spack/repos/builtin/packages/foam-extend/package.py [hunk 1]
```diff
@@ -114,187 +108,73 @@ class FoamExtend(Package):
...
+        run_env.set('FOAM_PROJECT_DIR', self.projectdir)
         run_env.set('WM_PROJECT_DIR', self.projectdir)
-
-    @property
-    def _canonical(self):
-        """Canonical name for this package and version"""
-        return 'foam-extend-{0}'.format(self.version.up_to(2))
+        for d in ['wmake', self.archbin]:  # bin already added automatically
+            run_env.prepend_path('PATH', join_path(self.projectdir, d))
+        run_env.set('MPI_BUFFER_SIZE', "20000000")
```
Comment (olesenm) [inline var/spack/repos/builtin/packages/foam-extend/package.py]: This is always set by the OpenFOAM environment scripts and seems to apply to all MPIs except SGI. Presumably this could be addressed if/when the setup_environment gets sorted.

File: var/spack/repos/builtin/packages/foam-extend/assets/spack-Allwmake [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+(
+    cd "$FOAM_LIBBIN" || exit 1
+    for i in $FOAM_MPI/lib*.so
+    do
+        [ -f $i ] && ln -sf $i "${i##*/}"
+    done
+)
+fi
+
+# -----------------------------------------------------------------------------
```
Comment (olesenm) [inline var/spack/repos/builtin/packages/foam-extend/assets/spack-Allwmake]: The files are almost identical, except that the handling of FOAM_INST_DIR has changed somewhat. The problem with from_sourcing_files seems to be unrelated to the installation directory. Maybe it has issues with some of the shell functions? I don't really know.

Comment (olesenm) [inline var/spack/repos/builtin/packages/foam-extend/assets/spack-Allwmake]: The from_sourcing_files appears to bomb out (ImportError can't find site) when using my system python (2.7.12), but seems to work if I make python a build requirement for openfoam.

Nonetheless, I still can't do much with the environment settings since they called before the files are actually installed.

Comment (nrichart) [inline var/spack/repos/builtin/packages/foam-extend/assets/spack-Allwmake]: How it was done before was by creating a `EnvironmentModifications` en applying it to the current environment, this can be done in the `install` function.

```
def get_openfoam_environment(self):
        return EnvironmentModifications.from_sourcing_files(
            join_path(self.stage.source_path, 'etc/bashrc'))
```
That can be used anywhere like
```
env_openfoam = self.get_openfoam_environment()
env_openfoam.apply_modifications()
```

Comment (nrichart) [inline var/spack/repos/builtin/packages/foam-extend/assets/spack-Allwmake]: But again by writing a `WMakePackage` this might be solved if the order of the stages can be a bit rearranged.  I have to ask @alalazo if it is doable. In this case What you put in common in `OpenfoamArch` could be inherited and make it easier to factorize code.

Comment (alalazo) [inline var/spack/repos/builtin/packages/foam-extend/assets/spack-Allwmake]: If it makes sense, writing a `WMakePackage` is absolutely doable. All you need to do is have a new module under the `build_system` folder and collect in the base class whatever can be reused. All the OpenFOAM forks (3 by now?) will inherit from `WMakePackage` and will contain only the part that are different from one to another.

File: var/spack/repos/builtin/packages/foam-extend/package.py [hunk 3]
```diff
@@ -111,189 +107,77 @@ class FoamExtend(Package):
     depends_on('scotch~metis+mpi', when='+ptscotch')
     depends_on('metis@5:',         when='+metis')
     depends_on('parmetis',         when='+parmetis')
-    depends_on('parmgridgen',      when='+parmgridgen')
+    # mgridgen is statically linked
+    depends_on('parmgridgen',      when='+parmgridgen', type='build')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/foam-extend/package.py]: Might want to keep this as the default: build/link. Otherwise Spack's compiler wrappers won't automatically add `-L/path/to/mgridgen/lib`.

File: var/spack/repos/builtin/packages/foam-extend/package.py [hunk 4]
```diff
@@ -111,189 +107,77 @@ class FoamExtend(Package):
...
+    def setup_dependent_environment(self, spack_env, run_env, dependent_spec):
+        """Provide location of the OpenFOAM project.
+        This is identical to the WM_PROJECT_DIR value, but we avoid that
+        variable since it would mask the normal OpenFOAM cleanup of
+        previous versions.
+        """
+        spack_env.set('FOAM_PROJECT_DIR', self.projectdir)
 
     @property
     def projectdir(self):
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/foam-extend/package.py]: Is this attribute useful for anything? Why not always use `prefix`? That's what every other package in Spack does.

File: var/spack/repos/builtin/packages/foam-extend/package.py (hunk 3 above)
Comment (olesenm) [inline var/spack/repos/builtin/packages/foam-extend/package.py]: I understand, but it will get linked in via the OpenFOAM build logic. In this case spack is reduced to feeding the OpenFOAM config with appropriate paths of where to find things and the OpenFOAM build manages the rest.

Comment (adamjstewart) [inline var/spack/repos/builtin/packages/foam-extend/package.py]: Is there any reason not to use the default dependency type? Just because OpenFOAM is special doesn't mean it doesn't link to that library.

File: var/spack/repos/builtin/packages/foam-extend/package.py (hunk 4 above)
Comment (olesenm) [inline var/spack/repos/builtin/packages/foam-extend/package.py]: Hi Adam,

For the top bit.
Yes I am an OpenFOAM developer (ESI-OpenCFD: www.openfoam.com).
Moving to another make system is fairly unlikely at this point. The hand-rolled `wmake`system is actually just a bunch of wrappers around a standard make, but obviously without all of the autotools or cmake configuration guessing. The main distinction though is probably that we have various lnInclude (symlinked include directory) with the headers for each major subsection. This means that the make file has the directory reference (via the `-I...` compiler option) without needing really long `#include ...` lines everywhere. If we go away from that, everything breaks and we have thousands of files...

OpenFOAM is indeed still a monster for using environment variables. The normal spack prefix is mostly okay, but I need to be able to build OpenFOAM with spack, but also be able to source the OpenFOAM environ and build depend modules without necessarily using spack.  For some (older) versions of OpenFOAM we some weird logic within OpenFOAM to find where the bashrc/cshrc files are located. I hope that we can make this even smoother in the future and drop this variable too.

Once the modules support fits better we should also be able to improve. If we can get this pull request merged in, any subsequent adjustments will presumably return the 'normal' range of 10-20 lines.

Cheers,
/mark

Comment (adamjstewart) [inline var/spack/repos/builtin/packages/foam-extend/package.py]: @olesenm Makes sense. I can only imagine how difficult it would be to make such a substantial change to the build system, especially now that OpenFOAM development has been split into 3 different branches: openfoam-com, openfoam-org, and foam-extend. I assume you primarily develop openfoam-com?

Also, thank you so much for taking the time to get these packages to build. I remember the last time one of our users requested OpenFOAM (org), I spent several weeks just trying to get it to build. I spent another several weeks trying to package it into Spack based on the foam-extend package, but had no luck. It's really nice when developers package their own software because I imagine you know what you're doing a lot better than I do.

**Comment by adamjstewart** [comment]
I'm just going to merge this. There is certainly an enormous amount of room for future improvement and ways that we can cut back on code that is duplicated across all 3 packages, but I don't want that to hold up @olesenm's work. It's a rare opportunity to see a developer interested in packaging their own software, and I don't want to scare off @olesenm or anyone else from contributing to Spack. God knows I would not want to have to write these packages myself.

Thanks @olesenm for all of your contributions! I'll take a look at #3650 and #3726 next.
